File: backend/dataset_builder/generate_projects.js
const axios = require('axios');
const fs = require('fs');
const path = require('path');
require('dotenv').config({ path: path.join(__dirname, '..', '.env') });

const KEY = process.env.GROQ_API_KEY;
const OUTPUT = path.join(__dirname, '..', 'data', 'projects.json');
const roleDB = JSON.parse(fs.readFileSync(path.join(__dirname, '..', 'data', 'role_skills_db.json'), 'utf8'));
const ALL_ROLES = Object.keys(roleDB);

let projects = {};
if (fs.existsSync(OUTPUT)) {
  projects = JSON.parse(fs.readFileSync(OUTPUT, 'utf8'));
  console.log(`Resuming: ${Object.keys(projects).length} roles done`);
}

const save = () => fs.writeFileSync(OUTPUT, JSON.stringify(projects, null, 2));
const sleep = ms => new Promise(r => setTimeout(r, ms));

async function ask(prompt) {
  const res = await axios.post('https://api.groq.com/openai/v1/chat/completions', {
    model: 'llama-3.3-70b-versatile',
    messages: [{ role: 'user', content: prompt }],
    max_tokens: 800,
    temperature: 0.3
  }, {
    headers: { 'Authorization': `Bearer ${KEY}`, 'Content-Type': 'application/json' },
    timeout: 30000
  });
  return res.data.choices[0].message.content.trim();
}

async function main() {
  const pending = ALL_ROLES.filter(r => !projects[r]);
  console.log(`${pending.length} roles to process`);
  for (let i = 0; i < pending.length; i += 3) {
    const batch = pending.slice(i, i + 3);
    const prompt = `2 portfolio projects per role for Indian students to get their first job.\nProject 1=Beginner(10-15hrs,free tools). Project 2=Intermediate(20-30hrs,free tools).\nEach: title, level, skills_demonstrated(array), hours_estimate, tool, how_to_share\nRoles: ${JSON.stringify(batch)}\nReply ONLY valid JSON no explanation: {"RoleName": [{"title":"...","level":"Beginner","skills_demonstrated":["Excel"],"hours_estimate":12,"tool":"MS Excel (free)","how_to_share":"Post screenshot on LinkedIn"}]}`;
    try {
      let raw = await ask(prompt);
      raw = raw.replace(/```json|```/g, '').trim();
      const result = JSON.parse(raw.slice(raw.indexOf('{'), raw.lastIndexOf('}') + 1));
      Object.assign(projects, result);
      console.log(`✅ Batch ${Math.floor(i/3)+1}/${Math.ceil(pending.length/3)} — ${batch.join(', ')}`);
    } catch(e) { console.warn(`⚠️ Batch failed: ${e.message}`); }
    save();
    await sleep(500);
  }
  console.log(`\nDone! projects.json → ${Object.keys(projects).length} roles`);
}
main().catch(console.error);

File: backend/dataset_builder/generate_projects.py
import requests
import json
import time
import os

# --- SMAART Dataset Builder: Portfolio Projects ---
# Generates 2 relevant portfolio projects for students per role 
# to ensure they have actionable tasks to close their skill gaps.

OPENROUTER_KEY = os.environ.get("OPENROUTER_API_KEY", "sk-or-your-key-here")
OUTPUT_FILE = os.path.join(os.path.dirname(__file__), '..', 'data', 'projects.json')

ROLE_DB_PATH = os.path.join(os.path.dirname(__file__), '..', 'data', 'role_skills_db.json')

if not os.path.exists(ROLE_DB_PATH):
    print(f"ERROR: {ROLE_DB_PATH} not found. Run generate_skills first.")
    exit(1)

with open(ROLE_DB_PATH) as f:
    role_db = json.load(f)
ALL_ROLES = list(role_db.keys())

# Resume Progress
projects = {}
if os.path.exists(OUTPUT_FILE):
    with open(OUTPUT_FILE) as f:
        try:
            projects = json.load(f)
            print(f"✅ Found {len(projects)} existing project mappings.")
        except:
            print("⚠️ Progress file corrupted, starting fresh.")

def save():
    with open(OUTPUT_FILE, 'w') as f:
        json.dump(projects, f, indent=2)

def call_mistral(prompt):
    res = requests.post(
        "https://openrouter.ai/api/v1/chat/completions",
        headers={
            "Authorization": f"Bearer {OPENROUTER_KEY}",
            "Content-Type": "application/json",
            "HTTP-Referer": "https://smaart.careers",
            "X-Title": "SMAART Career Platform"
        },
        json={
            "model": "mistralai/mistral-7b-instruct:free",
            "messages": [{"role": "user", "content": prompt}],
            "max_tokens": 800
        },
        timeout=30
    )
    return res.json()["choices"][0]["message"]["content"].strip()

# Batch roles by 3 to allow more tokens for project descriptions
for i in range(0, len(ALL_ROLES), 3):
    batch = [r for r in ALL_ROLES[i:i+3] if r not in projects]
    if not batch:
        continue

    prompt = f"""For each of these career roles, generate 2 portfolio projects that an Indian student can build to get their first job.

Project 1: Beginner level (10-15 hours, free tools only).
Project 2: Intermediate level (20-30 hours, free tools only).

For each project provide:
- title: short project name
- level: "Beginner" or "Intermediate"
- skills_demonstrated: list of 2-4 skill names from the role
- hours_estimate: number
- tool: free tool to use (e.g. MS Excel (free), VS Code)
- how_to_share: one-line advice on how to share this on LinkedIn/GitHub

Roles: {json.dumps(batch)}

Reply ONLY with valid JSON. Format:
{{"Financial Analyst": [{{"title": "...", "level": "Beginner", "skills_demonstrated": ["Excel"], "hours_estimate": 12, "tool": "MS Excel (free)", "how_to_share": "Screenshot on LinkedIn"}}]}}"""

    try:
        raw = call_mistral(prompt)
        raw = raw.replace("```json", "").replace("```", "").strip()
        start = raw.find("{")
        end = raw.rfind("}") + 1
        result = json.loads(raw[start:end])
        projects.update(result)
        print(f"✅ Processed Batch {i//3 + 1}/{len(ALL_ROLES)//3 + 1} — {list(result.keys())}")
        save()
        time.sleep(2)
    except Exception as e:
        print(f"⚠️ Failed market batch {i//3 + 1}: {str(e)}")
        time.sleep(5)

print(f"\n🏆 Generation Complete. Result stored in {OUTPUT_FILE}")
